File: src/main/java/gargoyle/calendar/util/load/Loaders.java
package gargoyle.calendar.util.load;

import gargoyle.calendar.util.resources.Resource;

import java.io.IOException;
import java.text.MessageFormat;
import java.util.LinkedHashMap;
import java.util.Map;

public class Loaders {

    private final Map<Class, Loader> loaders;

    protected Loaders() {
        loaders = new LinkedHashMap<>();
    }

    public Loaders(Map<Class, Loader> loaders) {
        this.loaders = new LinkedHashMap<>(loaders);
    }

    protected final <T> Loaders addLoader(Class<T> targetType, Loader<T> loader) {
        if (canLoad(targetType)) {
            throw new IllegalArgumentException(MessageFormat.format("loader {0} already exists", targetType));
        }
        loaders.put(targetType, loader);
        return this;
    }

    public final <T> boolean canLoad(Class<T> targetType) {
        return loaders.containsKey(targetType);
    }

    public final <T> T load(Class<T> targetType, Resource value) throws IOException {
        if (!loaders.containsKey(targetType)) throw new UnsupportedOperationException(MessageFormat.format(
                "loader {0} is not supported yet", targetType));
        Loader<T> loader = loaders.get(targetType);
        return loader.load(value);
    }

    @Override
    public final String toString() {
        return String.format("Loaders{loaders=%s}", loaders);
    }
}

File: cal-core/src/main/kotlin/gargoyle/calendar/util/load/Loaders.kt
package gargoyle.calendar.util.load

import gargoyle.calendar.util.resources.Resource
import java.text.MessageFormat
import kotlin.reflect.KClass

open class Loaders {
    private val loaders: MutableMap<KClass<*>, (Resource) -> Any>

    protected constructor() {
        loaders = LinkedHashMap()
    }

    constructor(loaders: Map<KClass<*>, (Resource) -> Any>) {
        this.loaders = LinkedHashMap(loaders)
    }

    protected fun <T : Any> addLoader(targetType: KClass<T>, loader: (Resource) -> T): Loaders {
        require(!canLoad(targetType)) { MessageFormat.format("loader {0} already exists", targetType) }
        loaders[targetType] = loader
        return this
    }

    fun <T : Any> canLoad(targetType: KClass<T>): Boolean = targetType in loaders

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> load(targetType: KClass<T>, value: Resource): T =
        (requireNotNull(loaders[targetType]) { "loader $targetType is not supported yet" } as (Resource) -> T)(value)

    override fun toString(): String = "Loaders{loaders=$loaders}"
}
